src/go/rdctl/pkg/reg/reg.go

```diff
@@ -0,0 +1,160 @@
+		if v.IsNil() {
+			return []string{}, true, nil
+		} else {
+			return convertToRegFormat(pathParts, v.Elem(), jsonTag)
+		}
+	case reflect.Slice, reflect.Array:
+		// Currently, all arrays in the options are arrays of strings
+		numValues := v.Len()
+		if numValues == 0 {
+			return nil, true, nil
+		}
+		arrayValues := make([]string, numValues)
+		for i := 0; i < numValues; i++ {
+			arrayValues[i] = v.Index(i).String()
+		}
+		return []string{fmt.Sprintf(`"%s"=hex(7):%s`, jsonTag, stringToMultiStringHexBytes(arrayValues))}, false, nil
+	case reflect.Map:
+		numValues := len(v.MapKeys())
+		if numValues == 0 {
+			return nil, true, nil
+		}
+		retLines := []string{"", fmt.Sprintf("[%s]", strings.Join(pathParts, "\\"))}
+		for _, key := range v.MapKeys() {
+			keyAsString := key.String()
+			innerLines, isEmpty, err := convertToRegFormat(append(pathParts, keyAsString), v.MapIndex(key), keyAsString)
+			if err != nil {
+				return nil, isEmpty, err
+			} else if !isEmpty {
+				retLines = append(retLines, innerLines...)
+			}
+		}
+		return retLines, false, nil
+	case reflect.Interface:
+		if v.IsNil() {
+			return nil, true, nil
+		}
+		return convertToRegFormat(pathParts, v.Elem(), jsonTag)
+	case reflect.Bool:
+		boolValue := map[bool]int{true: 1, false: 0}[v.Bool()]
+		return []string{fmt.Sprintf(`"%s"=dword:%d`, jsonTag, boolValue)}, false, nil
+	case reflect.Int, reflect.Int8, reflect.Int16,
+		reflect.Int32, reflect.Uint, reflect.Uint8, reflect.Uint16,
+		reflect.Uint32:
+		return []string{fmt.Sprintf(`"%s"=dword:%x`, jsonTag, v.Int())}, false, nil
+	case reflect.Int64, reflect.Uint64:
+		return []string{fmt.Sprintf(`"%s"=qword:%x`, jsonTag, v.Int())}, false, nil
+	case reflect.Float32, reflect.Float64:
+		return []string{fmt.Sprintf(`"%s"=dword:%x`, jsonTag, int(v.Float()))}, false, nil
+	case reflect.String:
+		return []string{fmt.Sprintf(`"%s"="%s"`, jsonTag, escape(v.String()))}, false, nil
+	}
+	return nil, true, fmt.Errorf("convertToRegFormat: don't know how to process kind: %s, (%T), value: %v for var %s\n", kind, v, v, jsonTag)
+}
+
+// Encode multi-stringSZ settings in comma-separated ucs2 little-endian bytes
+// e.g.=> ["abc", "def"] would be ucs-2-encoded as '61,00,62,00,63,00,00,00,64,00,65,00,66,00,00,00,00,00'
+// where a null 16-bit word (so two 00 bytes) separate each pair of words and
+// two null 16-bit words ("00 00 00 00") indicate the end of the list
+func stringToMultiStringHexBytes(values []string) string {
+	valueString := strings.Join(values, "\x00")
+	hex := utf16.Encode([]rune(valueString))
+	hexChars := make([]string, len(hex)*2)
+	for i, h := range hex {
+		s := fmt.Sprintf("%04x", h)
+		hexChars[2*i] = s[2:4]
+		hexChars[2*i+1] = s[0:2]
+	}
+	return strings.Join(hexChars, ",") + ",00,00,00,00"
+}
+
+// JsonToReg - convert the json settings to a reg file
+/**
+ * @param hiveType: "hklm" or "hkcu"
+ * @param profileType: "defaults" or "locked"
+ * @param settingsBodyAsJSON - options marshaled as JSON
```

Comment:
Maybe I misunderstood this comment, but after spending a bit of time on it I disagree.

The problem with this approach is that it makes writing tests onerous without a code-generation tool.

It's easy to both write and read a block of JSON code for a specific test, and verify the output. But if I have to create a structure, I'll have to assign each constant value to a variable, and then assign pointers to each variable that we care about. It's time-consuming to do that, and tedious for anyone to read the code.

The result is the same whether we do the unmarshaling in the caller or in the `JsonToReg` function, but overall
it's simpler and clearer to do the unmarshaling in the latter. 

Talking about generalizing the types of structures we can pass to `JsonToReg` is YAGNI right now. Right now I don't reflect arrays of strings correctly, getting a ucs2 encoding of `"<interface {} Value> <interface {} Value> <interface {} Value>"` instead of the expected `"abc" "def" "ghi"`, but that's because when I see an array in the reflector, I'm assuming it's an array of strings.

Also the generic converter now intermixes scalars and nested structs, which breaks the registry format, because all the scalar entries have to appear immediately after their containing struct's header. No point trying to figure out why this is happening unless and until we decide to generalize the converter.

Note that the unmarshaled struct will still need to have JSON tags in order to generate registry values.